torch/autograd/gradcheck.py

```diff
@@ -401,6 +457,15 @@ def not_reentrant_error(error_str=''):
         if out_is_complex and not reentrant_with_imag_grad_out:
             return not_reentrant_error(' (calculated using complex valued grad output)')
 
+        if check_batched_grad:
+            assert reentrant
```

Comment:
This will make gradcheck raise an error even when raise_exception=False
Should this be handled via the `fail_test` function like all other failures here?

Reply:
This assertion is intentional, the code to check batched grad assumes that backward is reentrant. If someone e.g. moved the batched grad check to before failing with the reentrant check, this will give a loud error.

Reply:
I'll add a comment, please let me know if that sounds reasonable

Reply:
I didn't mean to remove it, but just do `fail_test("vmap check is expected to fail when the backward is not reentrant, you should either fix the reentrant test or disable the vmap check if the backward is expected not to be reentrant")`.
The gradcheck function should never do a hard crash when `raise_exception=False`.

Reply:
The assert can never be triggered by users or authors of PyTorch tests. That is because above this line, we check for `reentrant` and do an early return if `reentrant` is True. If the assert gets triggered, then that means a developer who is modifying `gradcheck` has done something wrong.

This means that although there is an `assert` line, the gradcheck function (as long as it does not get modified) actually never does a hard crash when `raise_exception=False`.

Reply:
Sounds good. I didn't knew that we are currently early exiting when not reentrant.
Could you just add a comment next to this assert then mentioning this?
Thanks for the explanation!